### DISCUSSION TITLE
Using composition in c#

### DISCUSSION THREAD

**Discussion by cordasfilip** [dotnet/csharplang#4819] [Language Ideas]
Similar to how we can add static using to files add the ability to use this ability on regular fields in a class or a struct. So give the following code.

```
class Position
{
      public int X{get;set;}
      public int Y{get;set;}
}
class Shape
{
     using pubilc Postition Position;
     //The following method should explicitly know to use X and Y from the position variable
    public void Sum()=>X+Y; 
  
}
```
There should also be the ability to call methods with out conversion so if a method has a signature like 
`public static Sum(Position position)=>position.X + position.Y;` 
you should be able to call the method like so
` Sum(shape);` with out the need to do `Sum(shape.position)`
This feature should allow multiple using statements on properties like
```class Shape
{
     using pubilc Postition Position;
     using public Color Color;
    ...
}
```
This feature could also work in method parameters like
`public static Sum(using this Position position)=>X + Y;`
and be able to call it like so `shape.Sum()`
For more detailed explanation of this feature you can check out the video explanation from the Jai Language [here](https://www.youtube.com/watch?v=ZHqFrNyLlpA)

**Comment by cheong00** [comment]
Please explain why not just make Position _interface_ and then let Shape implements it.

Also note that multiple inheritance of class is explicitly rejected idea on C# language.

**Reply by cordasfilip** [reply]
Well interface is not an implementation just a definition you would still need to implement the interface. You could still have the Position be implemented from an interface this is just syntactic sugar you can already do this just requires using position argument a bunch. As in most cases the place where this sort of sugar could help you is refactoring because you don't need to change in many places if you decide to change for example if you want to have a different place you want to hold the property.
And again this has nothing to do with inheritance as it's just a sorter way to do something you can already do.

**Reply by cordasfilip** [reply]
A good place where I find this annoying is when I want to wrap for example a dictionary or a list with a custom one and I need to implement the interface and then just do a call to an underlying List/Dictionary in all the places I don't need extra logic. Like in the [following code](https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/Servers/Connections.Abstractions/src/ConnectionItems.cs) from Asp.net. You could remove all the useless calls to the list. And would be more performant(not much but hey) since no method will not call a new method to do something unnecessary.

**Reply by cheong00** [reply]
That sounds something that can be simplified by [default implementations in interface](https://devblogs.microsoft.com/dotnet/default-implementations-in-interfaces/).

**Reply by cordasfilip** [reply]
@cheong00 no it really is not, the idea of this is with memory management, It helps you split memory in a better way without refactoring, for example now you could separate X part and Y part into two different arrays and do faster loops depending on the need. It can utilize cash better and other things as well. I think people are already doing it this way instead of inheritance but this just makes it easier to implement.

**Comment by AliAlbarrak** [comment]
Seems like you are trying to achieve something similar to Kotlin delegation
https://kotlinlang.org/docs/delegation.html

**Reply by cheong00** [reply]
In Kotlin you still seems to need to declare the interface type on the class.

That sounds like [Extension Methods](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods) to me, which is available since .NET v3.0 for implementation of LINQ methods.

**Comment by gulshan** [comment]
#4772 similar suggestion by me.

**Reply by cordasfilip** [reply]
Hmm strange I didn't noticed this I really looked for this and was sure more people would like it.

**Comment by acaly** [comment]
I think this is a very specific usage. Another idea is to write your own source generator, which copies each methods/properties into the targe type with implementations calling the original ones. This shouldn't be very difficult to write.
